code/iaas/service-discovery/server/src/main/java/io/cattle/platform/servicediscovery/service/impl/ServiceDiscoveryServerImpl.java

```diff
@@ -0,0 +1,296 @@
+        Map<String, Object> launchConfigItems = new HashMap<>();
+        Set<Integer> volumesFromServices = new HashSet<>();
+        for (String key : data.keySet()) {
+            ServiceDiscoveryConfigItem item = ServiceDiscoveryConfigItem.getServiceConfigItemByRancherName(key);
+            if (item != null && item.isLaunchConfigItem()) {
+                // special handling for volumesFromService
+                if (item.getRancherName().equals(ServiceDiscoveryConfigItem.VOLUMESFROMSERVICE.getRancherName())) {
+                    List<Integer> serviceIds = (List<Integer>) data.get(key);
+                    for (Integer serviceId : serviceIds) {
+                        // get all instances for the service
+                        List<? extends ServiceExposeMap> serviceInstancesMap = mapDao.findNonRemoved(
+                                ServiceExposeMap.class,
+                                Service.class, serviceId);
+                        for (ServiceExposeMap map : serviceInstancesMap) {
+                            volumesFromServices.add(map.getInstanceId().intValue());
+                        }
+                    }
+                    continue;
+                }
+                launchConfigItems.put(key, data.get(key));
+            }
+        }
+
+        if (!volumesFromServices.isEmpty()) {
+            List<Integer> volumesFrom = (List<Integer>) launchConfigItems.get(ServiceDiscoveryConfigItem.VOLUMESFROM
+                    .getRancherName());
+            if (volumesFrom == null) {
+                volumesFrom = new ArrayList<Integer>();
+            }
+            volumesFromServices.removeAll(volumesFrom);
+            volumesFrom.addAll(volumesFromServices);
+            launchConfigItems.put(ServiceDiscoveryConfigItem.VOLUMESFROM.getRancherName(), volumesFrom);
+        }
+
+        return launchConfigItems;
+    }
+
+    private String getInstanceName(Instance instance) {
+        if (instance != null && instance.getRemoved() == null) {
+            return instance.getUuid();
+        } else {
+            return null;
+        }
+    }
+
+    @SuppressWarnings("unchecked")
+    private Map<String, Object> getServiceDataAsMap(Service service) {
+        Map<String, Object> originalData = new HashMap<>();
+        originalData.putAll(DataUtils.getFields(service));
+        Map<String, Object> data = new HashMap<>();
+        data.putAll((Map<? extends String, ? extends Object>) originalData
+                .get(ServiceDiscoveryConstants.FIELD_LAUNCH_CONFIG));
+        originalData.remove(ServiceDiscoveryConstants.FIELD_LAUNCH_CONFIG);
+        data.putAll(originalData);
+        return data;
+    }
+
+    @Override
+    public int[] getWeights(int size, int total) {
+        int[] weights = new int[size];
+        Integer sum = 0;
+        int i = 0;
+        while (true) {
+            int value = total / size;
+            if (sum + value > total) {
+                weights[i - 1] = total - sum + value;
```

Comment:
@sonchang not that at this point weights[i-1] already has the value set. If you just reset it to be "total-sum", you will lose this previously set value, and the final sum will be less than the total.